notifications/postgres/notificator.go

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package postgres
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	notificationStorage "github.com/Peripli/service-manager/notifications/postgres/storage"
+	"github.com/Peripli/service-manager/storage"
+
+	"github.com/lib/pq"
+
+	"github.com/Peripli/service-manager/pkg/types"
+
+	"github.com/Peripli/service-manager/pkg/log"
+
+	"github.com/Peripli/service-manager/notifications"
+	"github.com/Peripli/service-manager/pkg/web"
+)
+
+const (
+	postgresChannel             = "notifications"
+	invalidRevisionNumber int64 = -1
+)
+
+type consumers map[string][]notifications.NotificationQueue
+
+type notificator struct {
+	isRunning   bool
+	isListening bool
+
+	queueSize int
+
+	isRunningMutex  *sync.RWMutex
+	connectionMutex *sync.Mutex
+	connection      notificationStorage.NotificationConnection
+
+	consumersMutex *sync.RWMutex
+	consumers      consumers
+	storage        notificationStorage.NotificationStorage
+
+	ctx   context.Context
```

Comment:
While it would be better - Start and Register/Unregister are called in different places - one during bootstrap, the other during websocket request. As there is an optimization to start listening for new notifications only after there is a registered consumer we need the global context and not the one from the request.